vita3k/renderer/src/vulkan/pipeline_cache.cpp

```diff
@@ -0,0 +1,550 @@
+        if (vkvert->attribute_infos.find(attribute.regIndex) == vkvert->attribute_infos.end())
+            continue;
+
+        used_streams |= (1 << attribute.streamIndex);
+
+        const SceGxmVertexStream &stream = vertex_program.streams[attribute.streamIndex];
+        const SceGxmAttributeFormat attribute_format = static_cast<SceGxmAttributeFormat>(attribute.format);
+        shader::usse::AttributeInformation info = vkvert->attribute_infos.at(attribute.regIndex);
+
+        uint8_t component_count = attribute.componentCount;
+        vk::Format format;
+        if (info.regformat) {
+            const int comp_size = gxm::attribute_format_size(attribute_format);
+            component_count = (comp_size * component_count + 3) / 4;
+            // regformat attributes are int32
+            format = translate_attribute_format(SCE_GXM_ATTRIBUTE_FORMAT_UNTYPED, component_count, true, true);
+        } else {
+            format = translate_attribute_format(attribute_format, attribute.componentCount, info.is_integer, info.is_signed);
+        }
+
+        attr_descr.push_back(vk::VertexInputAttributeDescription{
+            .location = info.location(),
+            .binding = attribute.streamIndex,
+            .format = format,
+            .offset = attribute.offset });
+    }
+
+    for (unsigned int stream_index = 0; stream_index < SCE_GXM_MAX_VERTEX_STREAMS; stream_index++) {
+        if (!(used_streams & (1 << stream_index)))
+            continue;
+
+        const SceGxmVertexStream &stream = vertex_program.streams[stream_index];
+
+        const bool is_instanced = gxm::is_stream_instancing(static_cast<SceGxmIndexSource>(stream.indexSource));
+
+        binding_descr.push_back(vk::VertexInputBindingDescription{
+            .binding = stream_index,
+            .stride = stream.stride,
+            .inputRate = is_instanced ? vk::VertexInputRate::eInstance : vk::VertexInputRate::eVertex });
+    }
+
+    vk::PipelineVertexInputStateCreateInfo vertex_input{};
+    vertex_input.setVertexBindingDescriptions(binding_descr);
+    vertex_input.setVertexAttributeDescriptions(attr_descr);
+    return vertex_input;
+}
+
+static vk::StencilOpState convert_op_state(const GxmStencilStateOp &state) {
+    return vk::StencilOpState{
+        .failOp = translate_stencil_op(state.depth_fail),
+        .passOp = translate_stencil_op(state.depth_pass),
+        .depthFailOp = translate_stencil_op(state.depth_fail),
+        .compareOp = translate_stencil_func(state.func)
+    };
+}
+
+vk::Pipeline PipelineCache::retrieve_pipeline(VKContext &context, SceGxmPrimitiveType &type, MemState &mem) {
+    current_context = &context;
+    const GxmRecordState &record = context.record;
+    // get the hash of the current context
+    constexpr size_t record_pipeline_len = offsetof(GxmRecordState, vertex_streams);
+    uint64_t key = XXH_INLINE_XXH3_64bits(&record, record_pipeline_len);
```

Comment:
What a key to use. I love it. I think this can get interesting when compiler inserts padding though. (e.g. there is no standard way to clear the padding bytes in c++)

Reply:
You're right, but the record is always the same (context.record), so if there are paddings they should keep the same value while the emulator is running.